File: 백준/Silver/24482. 알고리즘 수업 － 깊이 우선 탐색 4/알고리즘 수업 － 깊이 우선 탐색 4.py
# 24482. 알고리즘 수업 - 깊이 우선 탐색 4

import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)

def dfs(start, depth):
    visited[start] = depth

    for j in graph[start]:
        if visited[j] == -1:
            dfs(j, depth+1)

N, M, R = map(int, input().split())
graph = [[] for _ in range(N+1)]
visited = [-1] * (N+1)

for _ in range(M):
    s, e = map(int, input().split())
    graph[s].append(e)
    graph[e].append(s)

for i in graph:
    i.sort(reverse=True)

dfs(R, 0)

for i in visited[1:]:
    print(i)

File: 백준/Silver/24482. 알고리즘 수업 － 깊이 우선 탐색 4/알고리즘 수업 － 깊이 우선 탐색 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
    static int depth;
    static int[] checked;
    static ArrayList<Integer>[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int start = Integer.parseInt(st.nextToken());

        checked = new int[n+1];
        arr = new ArrayList[n+1];

        for (int i = 1; i <=n; i++){
            arr[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());

            arr[s].add(e);
            arr[e].add(s);
        }

        for (int i = 1; i <= n; i++){
            Collections.sort(arr[i], Collections.reverseOrder());
        }

        Arrays.fill(checked, -1);
        dfs(start, depth);

        for (int i = 1; i < checked.length; i++){
            System.out.println(checked[i]);
        }
    }

    private static void dfs(int node, int d){
        checked[node] = d;
        for (int i : arr[node]){
            if (checked[i] == -1){
                dfs(i, d+1);
            }
        }
    }
}
